src/cc/arduino/plugins/wifi101/flashers/java/NinaFlasher.java

```diff
@@ -0,0 +1,139 @@
+ * invalidate any other reasons why the executable file might be covered by
+ * the GNU General Public License.
+ */
+package cc.arduino.plugins.wifi101.flashers.java;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.*;
+import java.util.List;
+import java.io.ByteArrayOutputStream;
+import java.io.*;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import org.apache.commons.lang3.StringUtils;
+import java.security.MessageDigest;
+
+import cc.arduino.plugins.wifi101.certs.WiFi101Certificate;
+import cc.arduino.plugins.wifi101.certs.WiFi101CertificateBundle;
+
+import cc.arduino.plugins.wifi101.flashers.Flasher;
+import javax.swing.JProgressBar;
+
+public class NinaFlasher extends Flasher {
+
+	public byte[] md5Checksum;
+
+	public NinaFlasher(String _modulename, String _version, String _filename, boolean _certavail, ArrayList<String> _compatibleBoard) {
+		super(_modulename, _version, _filename, _certavail, _compatibleBoard);
+	}
+
+	@Override
+	public void updateFirmware(String port) throws Exception {
+		FlasherSerialClient client = null;
+		try {
+			file = openFirmwareFile();
+			progress(10, "Connecting to programmer...");
+			client = new FlasherSerialClient();
+			client.open(port);
+			client.hello();
+			int maxPayload = client.getMaximumPayload();
+
+			byte[] fwData = this.getData();
+			int size = fwData.length;
+			int address = 0x00000000;
+			int written = 0;
+
+			progress(20, "Erasing target...");
+			client.eraseFlash(address, size);
+
+			while (written < size) {
+				progress(20 + written * 40 / size, "Programming " + size + " bytes ...");
+				int len = maxPayload;
+				if (written + len > size) {
+					len = size - written;
+				}
+				client.writeFlash(address, Arrays.copyOfRange(fwData, written, written + len));
+				written += len;
+				address += len;
+			}
+
+			progress(55, "Verifying...");
+
+			String filepath = file.getPath();
+			md5Checksum = getMD5Checksum(filepath);
+
+			address = 0x00000000;
+			byte[] md5rec = client.md5Flash(address, size);
+			progress(75, "Verifying...");
+			if (Arrays.equals(md5rec, md5Checksum)) {
+				progress(100, "Done!");
+			} else {
+				throw new Exception("Error different MD5 Checksum.");
+			}
+		} finally {
+			if (client != null) {
+				client.close();
+			}
+		}
+	}
+
+	@Override
+	public void uploadCertificates(String port, List<String> websites) throws Exception {
+		//TODO:to be implemented
+	}
+
+	//TODO: public WiFi101CertificateBundle createBundleFromWebsites to be implemented
+	/*public WiFi101CertificateBundle createBundleFromWebsites(List<String> websites) throws Exception {
+		//TODO:to be implemented
+	}*/
+
+	public static byte[] getMD5Checksum(String filename) throws Exception{
+		try {
```

Comment:
Can we calculate the MD5 from the data in `updateFirmware`, instead of reading the file twice?

Reply:
yes now the file is opened only one times when  the ```updateFirmware``` is called NINAFlasher.cpp:65, after in the MD5 function is used the results of getData call in NINAFlasher.cpp:72, it is ok in this way?

Reply:
yes it's fine. the `String filepath = file.getPath();` is confusing. Let' remove it.